### ISSUE TITLE
Choose damping function for vdW 

### ISSUE THREAD

**Issue by NikoOinonen** [Probe-Particle/ppafm#48]
It was noted in #35 that the stand-alone vdW calculation has some unusual term, which was used as a quick solution for damping the force to prevent instability. There exist better functions for the damping, e.g. https://manual.q-chem.com/5.2/Ch5.S7.SS2.html. We should choose one that has a good compromise between stability and speed.

**Comment by ProkopHapala** [comment]
I this notebook I 'm plotting the damping function.
https://colab.research.google.com/drive/1_JHrgwBzhr9bZa2MiWZp4twaplHGD9xx?usp=sharing

as you can se the current damping function for force `f = c6/( r**8 + 180*c6)`
is not so bad:
![plot](https://user-images.githubusercontent.com/3408568/203822621-163b16dc-01e1-4f61-a676-e354a6deda8f.png)

**Comment by ondrejkrejci** [comment]
Yes, it looks pretty promising.
Well, how would it be with one of the "published" damping?
If possible could you also extract some DFT density about the atom and show us the full difference for that. It would be really nice to compare these. Thank you

**Comment by ProkopHapala** [comment]
Or function with very little performance overhead which is very easily tuneable is this one

```
damp =   ( 1-(r/(Rij*1.2))**2 ) ; damp[damp<0]=0  
F = Eij*r*( -C6/(r**8 + cdamp*damp  )  )  
```

![download-2](https://user-images.githubusercontent.com/3408568/203830650-5cb25e29-b5e4-46ff-b7cc-b04747ce3a86.png)

**Comment by NikoOinonen** [comment]
> as you can se the current damping function for force `f = c6/( r**8 + 180*c6)` is not so bad: ![plot](https://user-images.githubusercontent.com/3408568/203822621-163b16dc-01e1-4f61-a676-e354a6deda8f.png)

I'm not that much of an expert on the different force fields, but that does not look very good to my eye. We mostly operate around the region where LJ has its minimum, and the damped curve has already ~50% difference compared to the undamped curve around this region. On the other hand, the `c6_dr2` in the previous post looks much better.

**Comment by ProkopHapala** [comment]
> > as you can se the current damping function for force `f = c6/( r**8 + 180*c6)` is not so bad: ![plot](https://user-images.githubusercontent.com/3408568/203822621-163b16dc-01e1-4f61-a676-e354a6deda8f.png)
> 
> I'm not that much of an expert on the different force fields, but that does not look very good to my eye. We mostly operate around the region where LJ has its minimum, and the damped curve has already ~50% difference compared to the undamped curve around this region. On the other hand, the `c6_dr2` in the previous post looks much better.

Hi, yes there is big difference form atractive part, but it match to L-J as a whole. My reasoning behind it was that I don't want to cause some unexpected over-binding as long as  don't know how to repulsive part (calculated form densnity overlap) looks like. Because otherwise the result is too sensitive on exact cancelation of the attratice and repulsive terms. So I wanted to keep C6/r^6 at long range, but make it irrelevant in the area where the bond-contrast is observed.

**Comment by ProkopHapala** [comment]
OK, so I'll work on it already today, otherwise I would not get to it. Sorry for missing the meeting in the morning.

**Comment by ProkopHapala** [comment]
@yakutovicha : OK, I started to work on it, but now I realized that I should have perhaps made a branch before I started to edit the code in order to pull-request later (?). So can I do it somehow easily now?

**Comment by yakutovicha** [comment]
@ProkopHapala, yes. You can branch-out at any point int time:

```bash
$ git checkout -b my-new-working-branch
```

And continue working there

**Comment by ProkopHapala** [comment]
Hi, so I did a bit more detailed study of the damping functions from . https://manual.q-chem.com/5.2/Ch5.S7.SS2.html and some other constructed by me. And I have several conclusions:

1) There are two meaningful ways how construct the damping:
a) multiply the by a damping function  ```E_damped(r) = E(r) * fdamp(r) ``` 
b) add some function to denominator e.g. ```E_damped = 1/( C6*r^6 + func(r) )```

Most often the method (a) is used, because it is more general, but it is less efficient and it may lead to numerical problems if undamped function goes to infinity or to zero. Method (b) can be always easily converted to method (a) by setting ```fdamp(r) = 1/(1+func(r))```

2) From the function used for damping vdW in DFT-D methods (  https://manual.q-chem.com/5.2/Ch5.S7.SS2.html ) I tested * the exponential damping ```fdamp_Exp(r)=  1 /( 1 + exp( -k * ( (r/R) - 1 ) ) ))``` 
* and the "Chai" function ```fdamp_Chai(r)=  E = 1./( 1. + A * (R/r)^12  )```. 
The Chai function is fine, but the exponetial damping I don't like, because it wiggles to much and does not damp the singularity at r=0, also it is costly to evalute exp function.

 ![Derivs_fdamp_exp](https://user-images.githubusercontent.com/3408568/218094704-2a283f97-d960-49ba-ac12-17d4bd86a109.png)

![Derivs_fdamp_Chai](https://user-images.githubusercontent.com/3408568/218094717-1f3b60fb-6752-46b3-892e-be4b25c6361f.png)

3) Dapmping by adding function radial spline ```R2func(r) = 1-(r/R)^2``` work fine if I add it directly to force like ``` F(r)=6*r*C6/( r^8 + A*R2func(r) ) ``` rather than to energy. But in that case we don't have simple analytical formula for potential energy (if we need it?).  If I add it to energy like  ``` E(r)=-C6/(r^6 + A*R2func(r) ) ``` then the derivative (i.e.fForce) is not smooth, which may prevent relaxation from convergence. Therefore I think it is worth it to use just a bit more complicated /costly radial spline     ```R4func(r) = (1-(r/R)^2)^2```

![Derivs_R2func](https://user-images.githubusercontent.com/3408568/218096632-0b9eb9af-1955-4b4a-9232-fdfd6d15311f.png)

![Derivs_R4func](https://user-images.githubusercontent.com/3408568/218096641-7c1451b0-a2b9-4737-92c6-ae386d52aa1f.png)

4) Basically the same result as with fdamp_Chai I can obtain by adding ```1/r^6 ``` to denominator ``` D(r) = C6/( r^6 + (R/r)^6 ) ```. The advantage over using  ```fdamp_Chai(r)=  E = 1./( 1. + A * (R/r)^12  )``` is that there is not so much numerical error due to cancelation inf/inf resp 0/0 when approaching r=0. 

![Derivs_invRn](https://user-images.githubusercontent.com/3408568/218097894-bd041642-8030-4419-99c9-ff19e5b04982.png)

**Summary:**

So in summary we have these 3 nicely behaving damping function. `fdamp_Chai` is basically the same as `+invR6`. Personally I like the most the R4func considering numerical speed and the fact that it has best compromise between localization (sharpness, finite support) and smoothness (C1 smooth).

![summary](https://user-images.githubusercontent.com/3408568/218098470-57beb8f9-1571-465e-852e-dd8f76b6ff83.png)

All functions can be examined here:

https://colab.research.google.com/drive/1_JHrgwBzhr9bZa2MiWZp4twaplHGD9xx?usp=sharing

**Comment by NikoOinonen** [comment]
> So in summary we have these 3 nicely behaving damping function. fdamp_Chai is basically the same as +invR6. Personally I like the most the R4func considering numerical speed and the fact that it has best compromise between localization (sharpness, finite support) and smoothness (C1-smooth).
> ![summary](https://user-images.githubusercontent.com/3408568/218098470-57beb8f9-1571-465e-852e-dd8f76b6ff83.png)

@ProkopHapala The `R4func` also has its minimum closer to the LJ minimum and has a repulsive part, which looks like in the force curve would add a somewhat significant additional bump to the repulsive interaction in some narrow range. Do you think this could be a problem?

**Comment by ProkopHapala** [comment]
So I implemented these damping function into the C++ code ( @NikoOinonen you know where exactly to put it in the OpenCL code? )

This script can plot all the versions of damped vdW 
https://github.com/Probe-Particle/ppafm/blob/prokop-damp-func/dev/test_vdwDaping.py

The results looks like this:

![VdW_C6](https://user-images.githubusercontent.com/3408568/218780151-f894d864-b804-4725-a7f5-6102cbc68421.png)

![VdW_R2](https://user-images.githubusercontent.com/3408568/218780344-7ccf73f5-ebde-40cf-8597-6985b5ad0e37.png)

![VdW_R4](https://user-images.githubusercontent.com/3408568/218780372-0a239ee7-6fa3-4eb8-804a-9eefba9047ef.png)

![VdW_invR4](https://user-images.githubusercontent.com/3408568/218780185-02c7ca1b-b73e-4067-872c-c50e270db93b.png)

![VdW_invR8](https://user-images.githubusercontent.com/3408568/218780239-0704ac0e-c964-41e7-bb53-ec61c9c0148f.png)

**Comment by ProkopHapala** [comment]
@yakutovicha I tried to link this issue with this pull-request but it did some mess
https://github.com/Probe-Particle/ppafm/pull/87

**Comment by yakutovicha** [comment]
@ProkopHapala, it is enough to put the following line in the topmost comment:

![Screenshot 2023-02-15 at 13 37 42](https://user-images.githubusercontent.com/13118485/219029636-80e8a735-b84b-4013-99ec-082cc9b1ee68.png)

I just did that and the issue is now connected. It will be closed automatically once the PR is merged.

**Comment by NikoOinonen** [comment]
> So I implemented these damping function into the C++ code ( @NikoOinonen you know where exactly to put it in the OpenCL code? )

@ProkopHapala I'm implementing the FDBM for OpenCL on this branch https://github.com/Probe-Particle/ppafm/tree/fdbm_ocl, and I already added there a function for calculating just the vdW https://github.com/Probe-Particle/ppafm/blob/db95811e2d51c2d1a368f4a6f24efdb670d11ce6/pyProbeParticle/ocl/field.py#L1374-L1375 but it's without any damping at the moment. I will add the damping functions there and merge it as a part of #35.
